docs/proposals/high-availability.md

```diff
@@ -0,0 +1,34 @@
+
+3. Stateless load-balanced Daemons: Stateless daemons, such as the apiserver, can simply load-balance across any number of servers that are currently running.  Their general availability can be continuously updated, or published, such that load balancing only occurs across active participants.  This aspect of HA is outside of the scope of *this* proposal because there is already a partial implementation in the apiserver.
+
+
+## Design Discussion Notes on Leader Election
+For a very simple example of proposed behavior see: 
+* https://github.com/rrati/etcd-ha
+* go get github.com/rrati/etcd-ha
+
+In HA, the apiserver will be a gateway to etcd. It will provide an api for becoming master, updating the master lease, and releasing the lease.  This api is daemon agnostic, so to become the master the client will need to provide the daemon type and the lease duration when attemping to become master.  The apiserver will attempt to create a key in etcd based on the daemon type that contains the client's hostname/ip and port information. This key will be created with a ttl from the lease duration provided in the request.  Failure to create this key means there is already a master of that daemon type, and the error from etcd will propigate to the client.  Successfully creating the key means the client making the request is the master.  When updating the lease, the apiserver will update the existing key with a new ttl.  The location in etcd for the HA keys is TBD.
+
+Leader election is first come, first serve.  The first daemon of a specific type to request leadership will become the master.  All other daemons of that type will fail until the current leader releases the lease or fails to update the lease within the expiration time.  On startup, all daemons should attempt to become master.  The daemon that succeeds is the master and should perform all functions of that daemon.  The daemons that fail to become the master should not perform any tasks and sleep for their lease duration and then attempt to become the master again.
+
+The daemon that becomes master should create a Go routine to manage the lease.  This process should be created with a channel that the main daemon process can use to release the master lease.  Otherwise, this process will update the lease and sleep, waiting for the next update time or notification to release the lease.  If there is a failure to update the lease, this process should force the entire daemon to exit.  Daemon exit is meant to prevent potential split-brain conditions.  Daemon restart is implied in this scenario, by either the init system (systemd), or possible watchdog processes.  (See Design Discussion Notes)
+
+## Options added to daemons with HA functionality
+Some command line options would be added to daemons that can do HA:
+
+* Lease Duration - How long a daemon can be master
+
+* Number of Missed Lease Updates - How many updates can be missed before the lease as the master is lost
```

Comment:
In theory the lease is invalid as soon as it times out.  It needs to be renewed before the timeout.
But that assumes some sort of synchronized clocks between the client and server to ensure that they both invalidate at the correct time - which you don't have. As mentioned above, this stuff is tricky :-)
I imagine that you'll need some assumptions about relative clock synchronization, or more usefully network latency (within x milliseconds of each other), and a safety margin (x milliseconds before my lease is due to expire, assume that it's already expired).  That way you err on the side of having no masters rather than accidentally having more than one master.